x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -82,48 +89,169 @@ func (s *websocketStream) FollowStream(ctx context.Context) error {
-			} else {
-				s.log.Errorw("failed to read websocket data", "error", err)
+			s.log.Errorw("encountered an error while closing the websocket connection", "error", err)
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.log.Debugw("context cancelled, closing websocket connection")
+			return ctx.Err()
+		default:
+			_, message, err := c.ReadMessage()
+			if err != nil {
+				s.metrics.errorsTotal.Inc()
+				if isRetryableError(err) {
+					s.log.Debugw("websocket connection encountered an error, attempting to reconnect...", "error", err)
+					// close the old connection and reconnect
+					if err := c.Close(); err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("encountered an error while closing the websocket connection", "error", err)
+					}
+					// since c is already a pointer, we can reassign it to the new connection and the defer func will still handle it
+					c, resp, err = connectWebSocket(ctx, s.cfg, url, s.log)
+					handleConnectionResponse(resp, s.metrics, s.log)
+					if err != nil {
+						s.metrics.errorsTotal.Inc()
+						s.log.Errorw("failed to reconnect websocket connection", "error", err)
+						return err
+					}
+				} else {
+					s.log.Errorw("failed to read websocket data", "error", err)
+					return err
+				}
+			}
+			s.metrics.receivedBytesTotal.Add(uint64(len(message)))
+			state["response"] = message
+			s.log.Debugw("received websocket message", logp.Namespace("websocket"), string(message))
+			err = s.process(ctx, state, s.cursor, s.now().In(time.UTC))
+			if err != nil {
+				s.metrics.errorsTotal.Inc()
+				s.log.Errorw("failed to process and publish data", "error", err)
+				return err
 			}
-			return err
 		}
-		s.metrics.receivedBytesTotal.Add(uint64(len(message)))
-		state["response"] = message
-		s.log.Debugw("received websocket message", logp.Namespace("websocket"), string(message))
-		err = s.process(ctx, state, s.cursor, s.now().In(time.UTC))
-		if err != nil {
-			s.metrics.errorsTotal.Inc()
-			s.log.Errorw("failed to process and publish data", "error", err)
-			return err
+	}
+}
+
+// isRetryableError checks if the error is retryable based on the error type.
+func isRetryableError(err error) bool {
+	// check for specific network errors
+	var netErr *net.OpError
+	if errors.As(err, &netErr) {
+		switch {
+		case netErr.Op == "dial" && netErr.Err.Error() == "i/o timeout",
+			netErr.Op == "read" && netErr.Err.Error() == "i/o timeout",
+			netErr.Op == "read" && netErr.Err.Error() == "connection reset by peer",
+			netErr.Op == "read" && netErr.Err.Error() == "connection refused",
+			netErr.Op == "read" && netErr.Err.Error() == "connection reset",
+			netErr.Op == "read" && netErr.Err.Error() == "connection closed":
+			return true
+		}
+	}
+
+	// check for specific websocket close errors
+	var closeErr *websocket.CloseError
+	if errors.As(err, &closeErr) {
+		switch closeErr.Code {
+		case websocket.CloseGoingAway,
```

Comment:
Is it possible to use https://pkg.go.dev/github.com/gorilla/websocket#IsCloseError here?

Reply:
**IsCloseError** will require us to pass all the possible codes in the func param which can seem a bit messy when we are dealing with so many types of codes. It can be used but the signature is messy compared to the switch statement.